traits/observers/_filtered_trait_observer.py

```diff
@@ -0,0 +1,186 @@
+
+    Parameters
+    ----------
+    notify : boolean
+        Whether to notify for changes.
+    filter : callable(str, CTrait) -> boolean
+        A callable that receives the name of a trait and the corresponding
+        trait definition. The returned boolean indicates whether the trait is
+        observed. In order to remove an existing observer with the equivalent
+        filter, the filter callables must compare equally.
+    """
+
+    def __init__(self, notify, filter):
+        self.notify = notify
+        self.filter = filter
+
+    def __hash__(self):
+        """ Return a hash of this object."""
+        return hash((type(self).__name__, self.notify, self.filter))
+
+    def __eq__(self, other):
+        """ Return true if this observer is equal to the given one."""
+        return (
+            type(self) is type(other)
+            and self.notify == other.notify
+            and self.filter == other.filter
+        )
+
+    def iter_observables(self, object):
+        """ Yield the instance traits matching the filter given. If trait
+        definition cannot be found on the object, raise an error.
+
+        Parameters
+        ----------
+        object: object
+            Object provided by the ``iter_objects`` methods from another
+            observers or directly by the user.
+
+        Yields
+        ------
+        IObservable
+
+        Raises
+        ------
+        ValueError
+            If the given object cannot be handled by this observer.
+        """
+        try:
+            name_to_ctrait = object.traits()
+        except AttributeError:
+            raise ValueError(
+                "Unable to obtain trait definitions from {!r}".format(object)
+            )
+
+        for name, ctrait in name_to_ctrait.items():
+            if self.filter(name, ctrait):
+                yield object._trait(name, 2)
+
+    def iter_objects(self, object):
+        """ Yield the values of the traits that match the given filter. The
+        values will be given to the next observer(s) following this one in an
+        ObserverGraph.
+
+        If a value has not been set as an instance attribute, i.e. absent in
+        ``object.__dict__``, this observer will skip it. This is to avoid
+        evaluating default initializers while adding observers. When the
+        default is defined, a change event will trigger the maintainer to
+        add/remove notifiers for the next observers.
+
+        Note that ``Undefined``, ``Uninitialized`` and ``None`` values are also
+        skipped, as they are unavoidable filled values that contain no further
+        attributes to be observed by any other observers.
+
+        Parameters
+        ----------
+        object: object
+            Object provided by the ``iter_objects`` methods from another
+            observers or directly by the user.
+
+        Yields
+        ------
+        value : object
+        """
+        try:
+            name_to_ctrait = object.traits()
```

Comment:
LBYL may be better here?, e.g. : `isinstance(object, CHasTraits)`

Reply:
I think the duck-typed way is better here.

Reply:
No problem.
Now I just recall this bug: https://github.com/enthought/traits/issues/674 that affects `traits` too. :/  Not sure I want this observer to know about that and try to workaround it.

Reply:
Agreed; that fix seems orthogonal.

I'll assign that bug to the 6.2 milestone, so that we don't forget about it; it would be good to resolve, but I don't want to expand this milestone any further than we have to.